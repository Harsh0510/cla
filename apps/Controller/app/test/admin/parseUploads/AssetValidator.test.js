const fs = require("fs");
const path = require("path");
const promisify = require("util").promisify;

const sax = require("sax");

const DefaultXmlParser = require("../../../core/admin/parseUploads/XmlParser");
const defaultHandlers = require("../../../core/admin/parseUploads/handlers");
const defaultGeneralErrorChecker = require("../../../core/admin/parseUploads/generalErrorChecker");
const defaultProductErrorCheck = require("../../../core/admin/parseUploads/productErrorCheck");

const AssetValidator = require("../../../core/admin/parseUploads/AssetValidator");

describe("check with set metods", () => {
	const func = (_) => 5;
	test("setting the meta data file fetcher", async () => {
		const ie = new AssetValidator();
		ie.setMetadataFileFetcher(func);
		expect(ie.metadataFileFetcher).toEqual(func);
	});

	test("setting the pdf binary file fetcher", async () => {
		const ie = new AssetValidator();
		ie.setPdfBinaryFileFetcher(func);
		expect(ie.pdfBinaryFileFetcher).toEqual(func);
	});

	test("setting the read stream creator", async () => {
		const ie = new AssetValidator();
		ie.setReadStreamCreator(func);
		expect(ie.readStreamCreator).toEqual(func);
	});

	test("setting the handlers", async () => {
		const ie = new AssetValidator();
		ie.setHandlers(func);
		expect(ie.handlers).toEqual(func);
	});

	test("setting the general error checker", async () => {
		const ie = new AssetValidator();
		ie.setGeneralErrorChecker(func);
		expect(ie.generalErrorChecker).toEqual(func);
	});

	test("setting the XML parser", async () => {
		const ie = new AssetValidator();
		ie.setXmlParser(func);
		expect(ie.xmlParser).toEqual(func);
	});

	test("setting the product error check function", async () => {
		const ie = new AssetValidator();
		ie.setProductErrorCheckFunction(func);
		expect(ie.productErrorCheck).toEqual(func);
	});

	test("setting the pdf page counter", async () => {
		const ie = new AssetValidator();
		ie.setPdfPageCounter(func);
		expect(ie.pdfPageCounter).toEqual(func);
	});
});

describe("process - with actual data", () => {
	test("executes successfully", async () => {
		const validator = new AssetValidator();
		const readDirPromise = promisify(fs.readdir);
		const dir = path.join(__dirname, "meta");
		const files = await readDirPromise(dir);
		files.sort();

		validator.setMetadataFileFetcher(async (_) => {
			const ret = [];
			for (const f of files) {
				if (f.match(/^[0-9]+\.xml$/) || f.match(/^magazine\.xml$/)) {
					ret.push(f);
				}
			}
			return ret;
		});
		validator.setPdfBinaryFileFetcher(async (product) => {
			if (!product.isbn13) {
				return null;
			}
			for (const f of files) {
				if (f.slice(-4) !== ".pdf") {
					continue;
				}
				if (f.indexOf(product.isbn13) !== -1) {
					return f;
				}
			}
		});
		validator.setReadStreamCreator((fp) => {
			return fs.createReadStream(path.join(dir, fp));
		});
		validator.setHandlers(defaultHandlers);
		validator.setGeneralErrorChecker(defaultGeneralErrorChecker);
		validator.setXmlParser(new DefaultXmlParser(sax));
		validator.setProductErrorCheckFunction(defaultProductErrorCheck);
		validator.setPdfPageCounter(async (_) => {
			return 0;
		});

		const results = await validator.process(dir);
		expect(results).toMatchSnapshot();
	});
	test("handler failure", async () => {
		const validator = new AssetValidator();
		const readDirPromise = promisify(fs.readdir);
		const dir = path.join(__dirname, "meta");
		const files = await readDirPromise(dir);
		files.sort();

		validator.setMetadataFileFetcher(async (_) => {
			const ret = [];
			for (const f of files) {
				if (f.match(/\btiny\.xml$/)) {
					ret.push(f);
				}
			}
			return ret;
		});
		validator.setPdfBinaryFileFetcher(async (product) => {
			if (!product.isbn13) {
				return null;
			}
			for (const f of files) {
				if (f.slice(-4) !== ".pdf") {
					continue;
				}
				if (f.indexOf(product.isbn13) !== -1) {
					return f;
				}
			}
		});
		validator.setReadStreamCreator((fp) => {
			return fs.createReadStream(path.join(dir, fp));
		});
		const handlers = defaultHandlers.slice(0);
		for (let i = 0, len = handlers.length; i < len; ++i) {
			if (handlers[i]._NAME_ === "authors") {
				handlers[i] = () => {
					throw new Error("MASSIVE ERROR");
				};
				handlers[i]._NAME_ = "authors";
				break;
			}
		}
		validator.setHandlers(handlers);
		validator.setGeneralErrorChecker(defaultGeneralErrorChecker);
		validator.setXmlParser(new DefaultXmlParser(sax));
		validator.setProductErrorCheckFunction(defaultProductErrorCheck);
		validator.setPdfPageCounter(async (_) => {
			return 0;
		});

		const results = await validator.process(dir);
		expect(results).toEqual([
			{
				errors: [],
				name: "tiny.xml",
				products: [
					{
						_errors: [expect.stringContaining("Internal handler error with handler 'authors': Error: MASSIVE ERROR")],
						collection: [],
						coverUris: [],
						edition: 1,
						educationalYearGroup: [],
						errors: [
							"PDF ISBN13 not found",
							"ISSN ID not found",
							expect.stringContaining("Internal handler error with handler 'authors': Error: MASSIVE ERROR"),
						],
						exam: [],
						examBoard: [],
						extent: 312,
						isbn13: "9781471892165",
						keyStage: [],
						level: [],
						language: ["eng"],
						pageOffsetArabic: 0,
						pageOffsetRoman: 0,
						publicationDate: 1498780800,
						recordReference: "CLAEP-00000002",
						scottishLevel: [],
						subjects: ["PN"],
						title: "TEST",
						parentAsset: {},
					},
				],
			},
		]);
	});

	test(`processing single XML ONIX file`, async () => {
		const validator = new AssetValidator();
		const readDirPromise = promisify(fs.readdir);
		const dir = path.join(__dirname, "meta");
		const files = await readDirPromise(dir);
		files.sort();

		validator.setMetadataFileFetcher(async (_) => {
			const ret = [];
			for (const f of files) {
				if (f.match(/^[0-9]+\.xml$/)) {
					ret.push(f);
				}
			}
			return ret;
		});
		validator.setPdfBinaryFileFetcher(async (product) => {
			if (!product.isbn13) {
				return null;
			}
			for (const f of files) {
				if (f.slice(-4) !== ".pdf") {
					continue;
				}
				if (f.indexOf(product.isbn13) !== -1) {
					return f;
				}
			}
		});
		validator.setReadStreamCreator((fp) => {
			return fs.createReadStream(path.join(dir, fp));
		});
		validator.setHandlers(defaultHandlers);
		validator.setGeneralErrorChecker(() => {
			throw new Error("Valid XML not found.");
		});
		validator.setXmlParser(new DefaultXmlParser(sax));
		validator.setProductErrorCheckFunction(defaultProductErrorCheck);
		validator.setPdfPageCounter(async (_) => {
			return 0;
		});

		const results = await validator.process(dir);
		expect(results).toEqual([
			{
				errors: [expect.stringContaining("Error: Valid XML not found.")],
				name: "1.xml",
				products: [],
			},
			{
				errors: [expect.stringContaining("Error: Valid XML not found.")],
				name: "10.xml",
				products: [],
			},
			{
				errors: [expect.stringContaining("Error: Valid XML not found.")],
				name: "11.xml",
				products: [],
			},
			{
				errors: [expect.stringContaining("Error: Valid XML not found.")],
				name: "4.xml",
				products: [],
			},
			{
				errors: [expect.stringContaining("Error: Valid XML not found.")],
				name: "5.xml",
				products: [],
			},
			{
				errors: [expect.stringContaining("Error: Valid XML not found.")],
				name: "6.xml",
				products: [],
			},
			{
				errors: [expect.stringContaining("Error: Valid XML not found.")],
				name: "7.xml",
				products: [],
			},
			{
				errors: [expect.stringContaining("Error: Valid XML not found.")],
				name: "8.xml",
				products: [],
			},
			{
				errors: [expect.stringContaining("Error: Valid XML not found.")],
				name: "9.xml",
				products: [],
			},
		]);
	});

	test("handler failure when unknown error", async () => {
		const validator = new AssetValidator();
		const readDirPromise = promisify(fs.readdir);
		const dir = path.join(__dirname, "meta");
		const files = await readDirPromise(dir);
		files.sort();

		validator.setMetadataFileFetcher(async (_) => {
			const ret = [];
			for (const f of files) {
				if (f.match(/\btiny\.xml$/)) {
					ret.push(f);
				}
			}
			return ret;
		});
		validator.setPdfBinaryFileFetcher(async (product) => {
			if (!product.isbn13) {
				return null;
			}
			for (const f of files) {
				if (f.slice(-4) !== ".pdf") {
					continue;
				}
				if (f.indexOf(product.isbn13) !== -1) {
					return f;
				}
			}
		});
		validator.setReadStreamCreator((fp) => {
			return fs.createReadStream(path.join(dir, fp));
		});
		const handlers = defaultHandlers.slice(0);
		handlers.push((product, productNode) => {
			throw new Error("Unknown error 1");
		});
		handlers.push((product, productNode) => {
			throw new Error("Unknown error 2");
		});
		for (let i = 0, len = handlers.length; i < len; ++i) {
			if (handlers[i]._NAME_ === "authors") {
				handlers[i] = () => {
					throw new Error("MASSIVE ERROR");
				};
				handlers[i]._NAME_ = "authors";
				break;
			}
		}

		validator.setHandlers(handlers);
		validator.setGeneralErrorChecker(defaultGeneralErrorChecker);
		validator.setXmlParser(new DefaultXmlParser(sax));
		validator.setProductErrorCheckFunction(defaultProductErrorCheck);
		validator.setPdfPageCounter(async (_) => {
			return 0;
		});

		const results = await validator.process(dir);

		expect(results[0].products[0]._errors.length).toEqual(3);
	});
});
