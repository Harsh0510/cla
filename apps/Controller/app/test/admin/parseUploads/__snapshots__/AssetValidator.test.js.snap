// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`process - with actual data executes successfully 1`] = `
Array [
  Object {
    "errors": Array [],
    "name": "1.xml",
    "products": Array [
      Object {
        "authors": Array [
          Object {
            "firstName": "David",
            "lastName": "Bowles",
            "roleCode": "A01",
            "sequenceNumber": 1,
          },
          Object {
            "firstName": "Brian",
            "lastName": "Jefferson",
            "roleCode": "A01",
            "sequenceNumber": 2,
          },
          Object {
            "firstName": "Eddie",
            "lastName": "Mullan",
            "roleCode": "A01",
            "sequenceNumber": 3,
          },
          Object {
            "firstName": "John",
            "lastName": "Rayneau",
            "roleCode": "A01",
            "sequenceNumber": 4,
          },
          Object {
            "firstName": "Mark",
            "lastName": "Rowland",
            "roleCode": "A01",
            "sequenceNumber": 5,
          },
          Object {
            "firstName": "Robert",
            "lastName": "Wagner",
            "roleCode": "A01",
            "sequenceNumber": 6,
          },
          Object {
            "firstName": "Paul",
            "lastName": "Williams",
            "roleCode": "A01",
            "sequenceNumber": 7,
          },
          Object {
            "firstName": "Garry",
            "lastName": "Wiseman",
            "roleCode": "A01",
            "sequenceNumber": 8,
          },
          Object {
            "firstName": "Katie",
            "lastName": "Wood",
            "roleCode": "A01",
            "sequenceNumber": 9,
          },
          Object {
            "firstName": "David",
            "lastName": "Baker",
            "roleCode": "B20",
            "sequenceNumber": 10,
          },
        ],
        "collection": Array [],
        "coverUris": Array [],
        "edition": 1,
        "educationalYearGroup": Array [],
        "errors": Array [
          "PDF ISBN13 not found",
          "ISSN ID not found",
          "Title not found",
        ],
        "exam": Array [],
        "examBoard": Array [],
        "extent": 349,
        "isbn13": "9780198412953",
        "keyStage": Array [],
        "language": Array [
          "eng",
        ],
        "level": Array [],
        "pageOffsetArabic": 0,
        "pageOffsetRoman": 0,
        "parentAsset": Object {},
        "publicationDate": 1491436800,
        "recordReference": "CLAEP-00000001",
        "scottishLevel": Array [],
        "subjects": Array [
          "PB",
        ],
      },
    ],
  },
  Object {
    "errors": Array [],
    "name": "10.xml",
    "products": Array [
      Object {
        "authors": Array [
          Object {
            "firstName": "Alyn G.",
            "lastName": "McFarland",
            "roleCode": "A01",
            "sequenceNumber": 1,
          },
          Object {
            "firstName": "Nora",
            "lastName": "Henry",
            "roleCode": "A01",
            "sequenceNumber": 2,
          },
        ],
        "collection": Array [],
        "coverUris": Array [],
        "edition": 1,
        "educationalYearGroup": Array [],
        "errors": Array [
          "PDF ISBN13 not found",
          "ISSN ID not found",
          "Title not found",
        ],
        "exam": Array [],
        "examBoard": Array [],
        "extent": 312,
        "isbn13": "9781471892165",
        "keyStage": Array [],
        "language": Array [
          "eng",
        ],
        "level": Array [],
        "pageOffsetArabic": 0,
        "pageOffsetRoman": 0,
        "parentAsset": Object {},
        "publicationDate": 1498780800,
        "recordReference": "CLAEP-00000002",
        "scottishLevel": Array [],
        "subjects": Array [
          "PN",
        ],
      },
      Object {
        "authors": Array [
          Object {
            "firstName": "Pauline",
            "lastName": "Lowrie",
            "roleCode": "A01",
            "sequenceNumber": 1,
          },
          Object {
            "firstName": "Mark",
            "lastName": "Smith",
            "roleCode": "A01",
            "sequenceNumber": 2,
          },
        ],
        "collection": Array [],
        "coverUris": Array [],
        "edition": 1,
        "educationalYearGroup": Array [],
        "errors": Array [
          "PDF ISBN13 not found",
          "ISSN ID not found",
          "Title not found",
        ],
        "exam": Array [],
        "examBoard": Array [],
        "extent": 288,
        "isbn13": "9781471807619",
        "keyStage": Array [],
        "language": Array [
          "eng",
        ],
        "level": Array [],
        "pageOffsetArabic": 0,
        "pageOffsetRoman": 0,
        "parentAsset": Object {},
        "publicationDate": 1427414400,
        "recordReference": "CLAEP-00000002",
        "scottishLevel": Array [],
        "subjects": Array [
          "PS",
        ],
      },
    ],
  },
  Object {
    "errors": Array [],
    "name": "11.xml",
    "products": Array [
      Object {
        "authors": Array [
          Object {
            "firstName": "Catherine",
            "lastName": "Casey",
            "roleCode": "A01",
            "sequenceNumber": 1,
          },
        ],
        "collection": Array [],
        "coverUris": Array [],
        "edition": 1,
        "educationalYearGroup": Array [],
        "errors": Array [
          "PDF ISBN13 not found",
          "ISSN ID not found",
          "Title not found",
        ],
        "exam": Array [],
        "examBoard": Array [],
        "extent": 80,
        "isbn13": "9781407160139",
        "keyStage": Array [],
        "language": Array [
          "eng",
        ],
        "level": Array [],
        "pageOffsetArabic": 0,
        "pageOffsetRoman": 0,
        "parentAsset": Object {},
        "publicationDate": 1472688000,
        "recordReference": "CLAEP-00000003",
        "scottishLevel": Array [],
        "subjects": Array [
          "YQC",
        ],
      },
    ],
  },
  Object {
    "errors": Array [],
    "name": "4.xml",
    "products": Array [
      Object {
        "authors": Array [
          Object {
            "firstName": "Tony",
            "lastName": "Gardiner",
            "roleCode": "A01",
            "sequenceNumber": 1,
          },
        ],
        "collection": Array [],
        "coverUris": Array [],
        "edition": 1,
        "educationalYearGroup": Array [],
        "errors": Array [
          "PDF ISBN13 not found",
          "ISSN ID not found",
          "Title not found",
        ],
        "exam": Array [],
        "examBoard": Array [],
        "extent": 144,
        "isbn13": "9781910504734",
        "keyStage": Array [],
        "language": Array [
          "eng",
        ],
        "level": Array [],
        "pageOffsetArabic": 0,
        "pageOffsetRoman": 0,
        "parentAsset": Object {},
        "publicationDate": 1514764800,
        "recordReference": "CLAEP-00000004",
        "scottishLevel": Array [],
        "subjects": Array [
          "PB",
        ],
      },
    ],
  },
  Object {
    "errors": Array [],
    "name": "5.xml",
    "products": Array [
      Object {
        "authors": Array [
          Object {
            "firstName": "Tony",
            "lastName": "Gardiner",
            "roleCode": "A01",
            "sequenceNumber": 1,
          },
        ],
        "collection": Array [],
        "coverUris": Array [],
        "edition": 1,
        "educationalYearGroup": Array [],
        "errors": Array [
          "PDF ISBN13 not found",
          "ISSN ID not found",
          "Title not found",
        ],
        "exam": Array [],
        "examBoard": Array [],
        "extent": 128,
        "isbn13": "9781910504727",
        "keyStage": Array [],
        "language": Array [
          "eng",
        ],
        "level": Array [],
        "pageOffsetArabic": 0,
        "pageOffsetRoman": 0,
        "parentAsset": Object {},
        "publicationDate": 1514764800,
        "recordReference": "CLAEP-00000005",
        "scottishLevel": Array [],
        "subjects": Array [
          "PB",
        ],
      },
    ],
  },
  Object {
    "errors": Array [],
    "name": "6.xml",
    "products": Array [
      Object {
        "authors": Array [
          Object {
            "firstName": "Tony",
            "lastName": "Gardiner",
            "roleCode": "A01",
            "sequenceNumber": 1,
          },
        ],
        "collection": Array [],
        "coverUris": Array [],
        "edition": 1,
        "educationalYearGroup": Array [],
        "errors": Array [
          "PDF ISBN13 not found",
          "ISSN ID not found",
          "Title not found",
        ],
        "exam": Array [],
        "examBoard": Array [],
        "extent": 104,
        "isbn13": "9781910504710",
        "keyStage": Array [],
        "language": Array [
          "eng",
        ],
        "level": Array [],
        "pageOffsetArabic": 0,
        "pageOffsetRoman": 0,
        "parentAsset": Object {},
        "publicationDate": 1514764800,
        "recordReference": "CLAEP-00000006",
        "scottishLevel": Array [],
        "subjects": Array [
          "PB",
        ],
      },
    ],
  },
  Object {
    "errors": Array [],
    "name": "7.xml",
    "products": Array [
      Object {
        "authors": Array [
          Object {
            "firstName": "Alyn G.",
            "lastName": "McFarland",
            "roleCode": "A01",
            "sequenceNumber": 1,
          },
          Object {
            "firstName": "Nora",
            "lastName": "Henry",
            "roleCode": "A01",
            "sequenceNumber": 2,
          },
        ],
        "collection": Array [],
        "coverUris": Array [],
        "edition": 1,
        "educationalYearGroup": Array [],
        "errors": Array [
          "PDF ISBN13 not found",
          "ISSN ID not found",
          "Title not found",
        ],
        "exam": Array [],
        "examBoard": Array [],
        "extent": 312,
        "isbn13": "9781471892165",
        "keyStage": Array [],
        "language": Array [
          "eng",
        ],
        "level": Array [],
        "pageOffsetArabic": 0,
        "pageOffsetRoman": 0,
        "parentAsset": Object {},
        "publicationDate": 1498780800,
        "recordReference": "CLAEP-00000002",
        "scottishLevel": Array [],
        "subjects": Array [
          "PN",
        ],
      },
      Object {
        "authors": Array [
          Object {
            "firstName": "Pauline",
            "lastName": "Lowrie",
            "roleCode": "A01",
            "sequenceNumber": 1,
          },
          Object {
            "firstName": "Mark",
            "lastName": "Smith",
            "roleCode": "A01",
            "sequenceNumber": 2,
          },
        ],
        "collection": Array [],
        "coverUris": Array [],
        "edition": 1,
        "educationalYearGroup": Array [],
        "errors": Array [
          "PDF ISBN13 not found",
          "ISSN ID not found",
          "Title not found",
        ],
        "exam": Array [],
        "examBoard": Array [],
        "extent": 288,
        "isbn13": "9781471807619",
        "keyStage": Array [],
        "language": Array [
          "eng",
        ],
        "level": Array [],
        "pageOffsetArabic": 0,
        "pageOffsetRoman": 0,
        "parentAsset": Object {},
        "publicationDate": 1427414400,
        "recordReference": "CLAEP-00000002",
        "scottishLevel": Array [],
        "subjects": Array [
          "PS",
        ],
      },
    ],
  },
  Object {
    "errors": Array [],
    "name": "8.xml",
    "products": Array [
      Object {
        "authors": Array [
          Object {
            "firstName": "Catherine",
            "lastName": "Casey",
            "roleCode": "A01",
            "sequenceNumber": 1,
          },
        ],
        "collection": Array [],
        "coverUris": Array [],
        "edition": 1,
        "educationalYearGroup": Array [],
        "errors": Array [
          "PDF ISBN13 not found",
          "ISSN ID not found",
          "Title not found",
        ],
        "exam": Array [],
        "examBoard": Array [],
        "extent": 80,
        "isbn13": "9781407160139",
        "keyStage": Array [],
        "language": Array [
          "eng",
        ],
        "level": Array [],
        "pageOffsetArabic": 0,
        "pageOffsetRoman": 0,
        "parentAsset": Object {},
        "publicationDate": 1472688000,
        "recordReference": "CLAEP-00000003",
        "scottishLevel": Array [],
        "subjects": Array [
          "YQC",
        ],
      },
    ],
  },
  Object {
    "errors": Array [],
    "name": "9.xml",
    "products": Array [
      Object {
        "authors": Array [
          Object {
            "firstName": "David",
            "lastName": "Bowles",
            "roleCode": "A01",
            "sequenceNumber": 1,
          },
          Object {
            "firstName": "Brian",
            "lastName": "Jefferson",
            "roleCode": "A01",
            "sequenceNumber": 2,
          },
          Object {
            "firstName": "Eddie",
            "lastName": "Mullan",
            "roleCode": "A01",
            "sequenceNumber": 3,
          },
          Object {
            "firstName": "John",
            "lastName": "Rayneau",
            "roleCode": "A01",
            "sequenceNumber": 4,
          },
          Object {
            "firstName": "Mark",
            "lastName": "Rowland",
            "roleCode": "A01",
            "sequenceNumber": 5,
          },
          Object {
            "firstName": "Robert",
            "lastName": "Wagner",
            "roleCode": "A01",
            "sequenceNumber": 6,
          },
          Object {
            "firstName": "Paul",
            "lastName": "Williams",
            "roleCode": "A01",
            "sequenceNumber": 7,
          },
          Object {
            "firstName": "Garry",
            "lastName": "Wiseman",
            "roleCode": "A01",
            "sequenceNumber": 8,
          },
          Object {
            "firstName": "Katie",
            "lastName": "Wood",
            "roleCode": "A01",
            "sequenceNumber": 9,
          },
          Object {
            "firstName": "David",
            "lastName": "Baker",
            "roleCode": "B20",
            "sequenceNumber": 10,
          },
        ],
        "collection": Array [],
        "coverUris": Array [],
        "edition": 1,
        "educationalYearGroup": Array [],
        "errors": Array [
          "PDF ISBN13 not found",
          "ISSN ID not found",
          "Title not found",
        ],
        "exam": Array [],
        "examBoard": Array [],
        "extent": 349,
        "isbn13": "9780198412953",
        "keyStage": Array [],
        "language": Array [
          "eng",
        ],
        "level": Array [],
        "pageOffsetArabic": 0,
        "pageOffsetRoman": 0,
        "parentAsset": Object {},
        "publicationDate": 1491436800,
        "recordReference": "CLAEP-00000001",
        "scottishLevel": Array [],
        "subjects": Array [
          "PB",
        ],
      },
    ],
  },
  Object {
    "errors": Array [],
    "name": "magazine.xml",
    "products": Array [
      Object {
        "authors": Array [
          Object {
            "firstName": "Rob",
            "lastName": "Attar",
            "roleCode": "B01",
            "sequenceNumber": 1,
          },
        ],
        "collection": Array [],
        "contentForm": "MI",
        "copyExcludedPages": Array [
          2,
          10,
          17,
          20,
          28,
          34,
          35,
          41,
          56,
          57,
          58,
          59,
          60,
          70,
          76,
          81,
          86,
          90,
          91,
          92,
          96,
        ],
        "coverUris": Array [
          "file://1469-8552-2016-01-199.png",
        ],
        "description": "<p>Many people make <b>new year's resolutions</b>: it's a tradition that goes back centuries, so it can be interesting to consider the resolutions that a few of history's best-known characters should have made as they began pivotal years in their lives. <i>BBC History</i> asked a panel of historians to come up with suggestions for the likes of <b>Anne Boleyn in 1536, Douglas Haig in 1916 and King Harold II in 1066</b>. You will find their ideas on page 29.</p><p>Harold II famously lost his life at the battle of Hastings, in a particularly unpleasant manner (if the legend is to be believed). Yet a violent death was nothing unusual in the Middle Ages, a period that was seemingly awash with blood. That, at least, is the popular conception of medieval life – but does this stereotype stand up to scrutiny? In the cover feature, Professor Hannah Skoda takes a fresh look at <b>medieval violence</b>. Turn to page 22 to find out more.</p><p>Finally, January 2016 saw the airing of the BBC's latest big-budget costume drama, an adaptation of <i>War and Peace</i>. Tolstoy's epic novel about the Russian war against Napoleon is regularly hailed – by those with the stamina to complete it – as one of the finest books ever written. On page 50, Dr Sarah Hudspith provides <b>a brief history of <i>War and Peace</i></b>.</p>",
        "edition": 1,
        "educationalYearGroup": Array [],
        "errors": Array [],
        "exam": Array [],
        "examBoard": Array [],
        "extent": 98,
        "isbn13": "1469-8552-2016-01-199",
        "issnId": "1469-8552-2016-01-199",
        "issueNumber": "January 2016",
        "keyStage": Array [],
        "language": Array [
          "eng",
        ],
        "level": Array [],
        "pageOffsetArabic": 0,
        "pageOffsetRoman": 0,
        "parentAsset": Object {
          "identifier": "1469-8552",
          "title": "BBC History Magazine",
        },
        "pdfIsbn13": "1469-8552-2016-01-199",
        "publicationDate": 1451606400,
        "publisher": "Immediate Media/BBC Studios",
        "recordReference": "uk.co.cla.1469-8552-2016-01-199",
        "scottishLevel": Array [],
        "subjects": Array [
          "HBG",
        ],
        "title": "BBC History Magazine - January 2016",
        "toc": "<ul> <li> <span class=\\"label\\">Anniversaries</span> <span class=\\"page\\">6</span> </li> <li> <span class=\\"label\\">History now</span> <span class=\\"page\\">11</span> </li> </ul>",
      },
    ],
  },
]
`;
